net: zperf: Add concurrency when doing upload tests

jukkar opened the pull request:
If user has enabled `CONFIG_ZPERF_SESSION_PER_THREAD`, then if user gives -a (async) option to upload command, then multiple uploads can be run simultaneously. Each upload will be run in a dedicated work queue. The work queue thread priority can be set by -t option.

Fixes #55963

jukkar committed:
net: zperf: Add concurrency when doing upload tests

If user has enabled CONFIG_ZPERF_SESSION_PER_THREAD, then if
user gives -a (async) option to upload command, then multiple
uploads can be run simultaneously. Each upload will be run
in a dedicated work queue. The work queue thread priority can
be set by -t option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/zperf/zperf_shell.c:
```diff
@@ -609,17 +658,32 @@ static void udp_upload_cb(enum zperf_status status,
+		shell_udp_upload_print_stats(sh, result, true);
+
+#ifdef CONFIG_ZPERF_SESSION_PER_THREAD
+		ses->in_progress = false;
```
I just wonder, shoudn't session state be updated by the respective work handler (`tcp_upload_async_work()`) when finished? In case someone uses zperf API instead of shell.

jukkar replied:
The `in_progress` is only used in zperf shell to determine what to print with `zperf jobs` command. I suppose we could set the variable in shell too. I can move the variable setting to there so we avoid setting these shell specific parameters in non-shell code.
I noticed also that we need to add test compilation to zperf sample for the new option which is not set by default.

jukkar replied:
Hmm, I just realized that we do not yet know the session in the shell when calling `tcp_upload_async_work()`, I think it is simpler if we leave the `in_progress` as is atm. If someone is not using shell, the variable is basically no-op and does no harm.

rlubos reviewed (dismissed):
Impressive!

jukkar committed:
samples: net: zperf: Add compilation of concurrent upload support

Make sure we compile test CONFIG_ZPERF_SESSION_PER_THREAD support.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: zperf: Add -w option to delay upload jobs

Use -w option to delay the startup of the upload job.
Then when ready, do "zperf jobs start" to launch all upload
sessions at the same time.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Added `zperf jobs start` command that can start all the waiting upload sessions at the same time. The upload session must be started with `-w` option in this case.

pdgendt commented on subsys/net/lib/zperf/zperf_common.c:
```diff
@@ -220,20 +278,52 @@ uint32_t zperf_packet_duration(uint32_t packet_size, uint32_t rate_in_kbps)
+#if defined(CONFIG_ZPERF_SESSION_PER_THREAD)
+	k_work_submit_to_queue(zperf_work_q[proto * SESSION_INDEX + session_id].queue, work);
+#else
 	k_work_submit_to_queue(&zperf_work_q, work);
```
```suggestion
	ARG_UNUSED(proto);
	ARG_UNUSED(session_id);

	k_work_submit_to_queue(&zperf_work_q, work);
```

jukkar commented:
* Rebased to main and resolved merge conflicts
* Updated according to comments
